### PR TITLE
1 )Ansible Improvements to provide the first class improvement to CLI…

### PR DESCRIPTION
**Pull request by hililh**
1) Ansible Improvements to provide the first class improvement to CLI option. 
2) Adding L0 test for ansible

### CODE REVIEW

File: Extensions/Ansible/Src/Tasks/Ansible/task.json [hunk 1]
```diff
@@ -56,69 +60,175 @@
...
+            "defaultValue": "agentMachine",
             "options": {
-                "commands": "Commands",
-                "script": "Shell script"
+                "agentMachine": "Agent machine",
+                "ansibleMachine": "Ansible machine"
             },
-            "helpMarkDown": "Choose to either run shell commands or a shell script on the remote machine."
+            "required": true,
+            "helpMarkDown": "Select agent machine, if playbook is located on automation agent. In this case playbook is copied to ansible machine. Otherwise, select ansible machine.",
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/task.json]: Lets have help text more verbose. I think we should fell free to use multiple lines to explain each option

File: Extensions/Ansible/Src/Tasks/Ansible/task.json [hunk 2]
```diff
@@ -56,69 +60,175 @@
...
+            "name": "playbookPathAnsibleMachine",
             "type": "string",
-            "label": "Arguments",
+            "label": "Path",
             "defaultValue": "",
-            "required": false,
-            "visibleRule": "cliRunOptions = script",
-            "helpMarkDown": "Arguments to pass to the shell script."
+            "required": true,
+            "helpMarkDown": "Path of the playbook files on execute on the ansible machine.",
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/task.json]: path of the playbook file  to execute

File: Extensions/Ansible/Src/vss-extension.json [hunk 3]
```diff
@@ -2,10 +2,10 @@
   "manifestVersion": 1.0,
   "id": "vss-services-ansible",
   "name": "Ansible",
-  "publisher": "ms-devlabs",
+  "publisher": "ms-vscs-rm",
```
Comment (sachinma) [inline Extensions/Ansible/Src/vss-extension.json]: why this change

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+            if (sshClientConnection) {
+
+                if(playbookSource == 'agentMachine') {
+                    
+                    let playbookRoot: string = tl.getInput('playbookRoot', true);
+                    let playbook: string = tl.getInput('playbookPathAgentMachine', true);
+                    
+                    let remotePlaybookRoot = '/tmp/' + path.basename(playbookRoot);
+                    let remotePlaybookRootPath = '"' + remotePlaybookRoot + '"';
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: why do you need to put it in quotes

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+                    tl.debug('ansiblePlaybookRootPath = ' + remotePlaybookRootPath);
+
+                    let scpConfig = sshConfig;
+                    scpConfig.path = remotePlaybookRoot;
+                    tl.debug('Copying playbook to ansible machine.');
+                    await ansibleUtils.copyFileToRemoteMachine(playbookRoot, scpConfig);
+
+                    let remotePlaybookPath = remotePlaybookRoot + "/" + path.basename(playbook);
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: Is there no path.join like function to join the paths

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+                    remoteCommand = remoteCommand.concat(" " + remotePlaybookPath)
+
+                    cleanupCmd.push('rm -rf ' + remotePlaybookRoot);
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: are you doing this before copy as well. I think that might be good to do in addition to this

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+                        remoteCommand = remoteCommand.concat(' -i ' + remoteInventory);
+
+                        cleanupCmd.push('rm -f ' + remoteInventory);
+
+                    } else if(inventoryFileSource == 'ansibleMachine') {
+
+                        let inventoryFile: string = tl.getInput('inventoryFileAnsibleMachine', true);
+                        tl.debug('InventoryFile = ' + inventoryFile);
+                        remoteCommand = remoteCommand.concat(' -i ' + inventoryFile);
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: consider moving this out of both if and else clause and both if and else only set inventoryFile

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+                    tl.debug('RemoteInventoryPath = ' + remoteInventoryPath);
+
+                    let inventoryCmd: string = 'echo ' + '"' + content + '"' + ' > ' + remoteInventory;
+                    await ansibleUtils.runCommandOnRemoteMachine(inventoryCmd, sshClientConnection);
+
+                    let dynamicInventory: boolean = tl.getBoolInput('inventoryInlineDynamic', true);
+                    
+                    if(dynamicInventory == true) {
+                        await ansibleUtils.runCommandOnRemoteMachine('chmod +x ' + remoteInventory, sshClientConnection);
+                    }
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: dont we need to add a flag for remote inventory

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts (hunk 4 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: To highlight the debug text. This path I am printing for debug

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts (hunk 6 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: I am not doing this before copy because it will copy will override, if the file is already present. This is as good as removing the old and pasting the new one. So from my understanding running the clean-up once after the copy should be sufficient

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts (hunk 7 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: ok

File: Extensions/Ansible/Src/vss-extension.json (hunk 3 above)
Comment (hililh) [inline Extensions/Ansible/Src/vss-extension.json]: Usha want it to publish as a Microsoft and not as MS devlabs

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts (hunk 5 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: We cannot use path.join()
agent could be running on windows machine, in this case path would join using back slash. where as this command needs to be run on Linux which require forward slash.
If the agent is on Linux machine then we could have used path.join

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts (hunk 8 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: doesn't required

File: Extensions/Ansible/Src/Tasks/Ansible/task.json (hunk 1 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/task.json]: I feel all this scenario should be captured on Readme file in verbose mode.
Will add it there

File: Extensions/Ansible/Src/Tasks/Ansible/task.json (hunk 2 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/task.json]: done

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts (hunk 4 above)
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: Consider not  creating var for it in that case

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts (hunk 5 above)
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: makes sense

File: Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts (hunk 6 above)
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/ansibleCommandLineInterface.ts]: But what if there are extra files

**Review by sachinma — APPROVED** [review]
